doc/devguide/extending/app-layer/app-layer-frames.rst

```diff
@@ -0,0 +1,227 @@
+
+- ``hdr``: used for the record header portion
+- ``data``: is used for the record data portion
+- ``pdu``: unless documented otherwise, means the whole record, comprising ``hdr`` and ``data``
+- ``request``: a message from a client to a server
+- ``response``: a message from a server to a client
+
+Once that is settled, the basic steps are:
+
+- create an enum with the frame types;
+- identify the parsing function(s) where application layer records are parsed;
+- identify the correct moment to register the frames (when the input stream is being parsed into records);
+- use the Frame API calls directly or build upon them and use your functions to register the frames;
+- register the relevant frame callbacks when registering the parser.
+
+Once these are done, you can enable frame eve-output to confirm that your frames are being properly registered. It is important to notice that some hard coded limits could influence what you see on the logs (max size of log output; type of logging for the payload, cf. https://redmine.openinfosecfoundation.org/issues/4988).
+
+If all the steps are successfully followed, you should be able to write a rule using the *frame* keyword and the frame types you have registered with the application layer parser.
+
+Using the *SMB* parser as example, before frame support, a rule would look like::
+
+    alert tcp ... flow:to_server; content:"|ff|SMB"; content:"some smb 1 issue";
+
+With frame support, one is able to do::
+
+    alert smb ... flow:to_server; frame:smb1.data; content:"some smb 1 issue";
+
+Implementation & API Callbacks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Rust
+----
+
+The frame types are defined as an enum. In Rust, derive from the ``AppLayerFrameType``:
+
+.. literalinclude:: ../../../../rust/src/smb/smb.rs
+    :caption: rust/src/smb/smb.rs
+    :language: rust
+    :start-at: #[derive(AppLayerFrameType)]
+    :lines: 1-15
+
+How some frames are registered in the `SMB <https://github.com/OISF/suricata/blob/master/rust/src/smb/smb.rs#L1383>`_ parser:
+
+.. literalinclude:: ../../../../rust/src/smb/smb.rs
+    :caption: rust/src/smb/smb.rs
+    :language: rust
+    :start-at: pub fn parse_tcp_data_ts<'b>
+    :end-at: nbss_hdr.data, nbss_hdr.length as i64);
+    :lines: 1-3, 56-74
+    :dedent: 4
+
+These are the frame registration functions highlighted above:
+
+.. literalinclude:: ../../../../rust/src/smb/smb.rs
+    :caption: rust/src/smb/smb.rs
+    :language: rust
+    :start-at: fn add_nbss_ts_frames(&mut
+    :end-before: fn add_smb1_ts_hdr_data
+    :dedent: 4
+
+Frame registering with the `telnet <https://github.com/OISF/suricata/blob/master/rust/src/telnet/telnet.rs#L171>`_ parser:
+
+.. _telnet-snippet:
+
+.. literalinclude:: ../../../../rust/src/telnet/telnet.rs
+    :caption: rust/src/telnet/telnet.rs
+    :language: rust
+    :start-at: fn parse_request
+    :end-at: TelnetFrameType::Data as u8)
```

Comment:
I reformatted some code and these includes started to fail.  I wonder if we should drop magic start/stop comments to match on instead? It would also be a clue in the code that modifying the code might need modification of the documentation.

Reply:
If we're ok with adding those, this is indeed better. It is how these options are usually used, I reckon. 

I do like the idea of having the failure because of that: it tells us that we may need to update the documentation...